File: apps/pwa/src/server/api/get_musicbill_list.ts
import { prefixServerOrigin } from '@/global_states/server';
import { request } from '..';

interface User {
  id: string;
  nickname: string;
  avatar: string;
}

type Response = {
  id: string;
  cover: string;
  name: string;
  public: boolean;
  createTimestamp: number;
  owner: User;
  sharedUserList: (User & {
    accepted: boolean;
  })[];
}[];

async function getMusicbillList() {
  const musicbillList = await request<Response>({
    path: '/api/musicbill_list',
    withToken: true,
  });
  return musicbillList.map((mb) => ({
    ...mb,
    cover: prefixServerOrigin(mb.cover),
    owner: {
      ...mb.owner,
      avatar: prefixServerOrigin(mb.owner.avatar),
    },
    sharedUserList: mb.sharedUserList.map((u) => ({
      ...u,
      avatar: prefixServerOrigin(u.avatar),
    })),
  }));
}

export default getMusicbillList;

File: apps/flutter/lib/server/api/get_musicbill_list.dart
import '../../utils/prefix_server_origin.dart';
import '../request.dart';

class Musicbill {
  final String id;
  final String name;
  final String? cover;
  final bool isPublic;
  final bool isShared;

  Musicbill({
    required this.id,
    required this.name,
    required this.cover,
    required this.isPublic,
    required this.isShared,
  });

  factory Musicbill.fromJson(Map<String, dynamic> json) => Musicbill(
    id: json['id'],
    name: json['name'],
    cover: prefixServerOrigin(json['cover']),
    isPublic: json['public'] == true,
    isShared: (json['sharedUserList'] as List<dynamic>? ?? const []).isNotEmpty,
  );
}

Future<List<Musicbill>> getMusicbillList() async {
  final responseData = await httpGet(
    path: "/api/musicbill_list",
    withToken: true,
  );
  return (responseData as List<dynamic>)
      .map((json) => Musicbill.fromJson(json))
      .toList();
}
